### PR TITLE
feat!: add transaction policies support

### PR DESCRIPTION
**Pull request by hal3e**
## Description
co-developed with: @IGI-111 and @xgreenx

This PR adds support for transaction policies. 

What was done:
- bump `fuel-vm` to `0.43.1`
- bump `fuels-rs` to `0.53.0`
- update the `std-lib` to handle new `GTF` codes and the transaction policies
- update all test

BREAKING CHANGE:
- removed `TxParameters` in favor of `TxPolicies`
- changed `gtf` opcodes 
- removed `Mint` from `forc-tx` and cli

## Checklist
- [ ] I have linked to any relevant issues.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

File: forc-plugins/forc-tx/src/lib.rs [hunk 1]
```diff
@@ -621,20 +635,19 @@ impl TryFrom<Create> for fuel_tx::Create {
             .into_iter()
             .map(|s| fuel_tx::Witness::from(s.as_bytes()))
             .collect();
+
+        let mut policies = Policies::default().with_maturity(create.maturity.maturity.into());
+        policies.set(PolicyType::GasPrice, create.gas.price);
```
Comment (xgreenx) [inline forc-plugins/forc-tx/src/lib.rs:640]:
```suggestion
        let mut policies = Policies::default().with_maturity(create.maturity.maturity.into()).with_gas_price( create.gas.price);
```

Also, we don't need to set `maturity` if it is zero.

File: test/src/sdk-harness/Cargo.toml [hunk 2]
```diff
@@ -36,5 +34,8 @@ path = "test_projects/harness.rs"
 
 [workspace]
 
-#[patch."https://github.com/FuelLabs/fuels-rs.git"]
-#fuels = { path = "/home/xunilrj/github/fuels-rs/packages/fuels" }
+[patch.crates-io]
+fuel-core = { git = "https://github.com/FuelLabs/fuel-core", branch = "master" }
+fuel-core-client = { git = "https://github.com/FuelLabs/fuel-core", branch = "master" }
```
Comment (xgreenx) [inline test/src/sdk-harness/Cargo.toml:39]: You can use `v0.21.0-rc.1` now=)

File: test/src/e2e_vm_tests/mod.rs [hunk 3]
```diff
@@ -173,6 +173,8 @@ impl TestContext {
                                 TestResult::ReturnData(data)
                             }
                             ProgramState::Revert(v) => TestResult::Revert(v),
+                            ProgramState::RunProgram(_) => todo!(), // TODO: @hal3e fix TestResult
```
Comment (hal3e) [inline test/src/e2e_vm_tests/mod.rs:176]: Can someone please have a look at this part of the code ?

File: forc-plugins/forc-tx/src/lib.rs [hunk 4]
```diff
@@ -62,10 +67,19 @@ pub struct Create {
 pub struct Mint {
     /// The location of the `Mint` transaction in the block.
     #[clap(long)]
-    pub tx_ptr: fuel_tx::TxPointer,
-    // Outputs must follow all other arguments and are parsed separately.
+    pub tx_ptr: fuel_tx::TxPointer, // TODO: @hal3e check if the `clap` settings are ok
```
Comment (hal3e) [inline forc-plugins/forc-tx/src/lib.rs:70]: The `Mint` transaction changed. Can someone please have a look at this part of the code ?

File: forc-plugins/forc-client/src/util/gas.rs [hunk 5]
```diff
@@ -17,6 +17,7 @@ pub fn get_gas_limit(gas: &Gas, chain_info: ChainInfo) -> u64 {
     if let Some(gas_limit) = gas.limit {
         gas_limit
     } else {
-        chain_info.consensus_parameters.max_gas_per_tx
+        //TODO: @hal3e estimate real `gas_used`
```
Comment (hal3e) [inline forc-plugins/forc-client/src/util/gas.rs:20]: Setting the script gas limit to `max_gas_per_tx` leads to a validation error. In the rust-sdk we dry run the `tx` first to get `gas_used` and set this as the limit. @xgreenx what do you recommend we do here ?

File: test/src/e2e_vm_tests/test_programs/should_pass/language/far_jumps/many_blobs/src/main.sw [hunk 6]
```diff
@@ -1,58 +1,60 @@
...
-            }
-            333
-        } else {
-            asm() {
-                blob i530000;
-            }
-            444
-        }
-    }
+//TODO: @hal3e figure out why this is failing now
```
Comment (hal3e) [inline test/src/e2e_vm_tests/test_programs/should_pass/language/far_jumps/many_blobs/src/main.sw:4]: This test does not work. I need some help to debug it.

File: test/src/e2e_vm_tests/test_programs/should_pass/stdlib/u256_div_test/src/main.sw [hunk 7]
```diff
@@ -14,13 +14,14 @@ fn main() -> bool {
...
-    let mut dividend = U256::from((0, 0, u64::max(), 1));
-    let mut res = dividend / max_u64;
+    let dividend = U256::from((0, 0, u64::max(), 1));
+    let res = dividend / max_u64;
     assert(res == U256::from((0, 0, 1, 0)));
 
-    dividend = U256::from((u64::max(), 0, 0, 0));
-    let mut res = dividend / max_u64;
-    assert(res == U256::from((1, 0, 0, 0)));
+    //TODO: @hal3e figure out why this is failing now
```
Comment (hal3e) [inline test/src/e2e_vm_tests/test_programs/should_pass/stdlib/u256_div_test/src/main.sw:21]: This part of the test does not work. I need some help to debug it.

File: forc-plugins/forc-client/src/util/gas.rs (hunk 5 above)
Comment (Torres-ssf) [reply forc-plugins/forc-client/src/util/gas.rs:20]: @hal3e Yeah, I was running into the same error on the TS SDK. 

Then I started doing exactly like you; a `dryRun` call then I get the `gas_used` and set it to `gas_limit`. 

However the `gas_limit` needs to be set on a `dryRun` call as well. I wonder what is the correct max value.

In the meanwhile I am using `max_gas_per_tx / 2`

File: test/src/e2e_vm_tests/test_programs/should_pass/stdlib/u256_div_test/src/main.sw (hunk 7 above)
Comment (xunilrj) [reply test/src/e2e_vm_tests/test_programs/should_pass/stdlib/u256_div_test/src/main.sw:21]: I uncommented this and the test is working.

Comment (xunilrj) [reply test/src/e2e_vm_tests/test_programs/should_pass/stdlib/u256_div_test/src/main.sw]: Done.

File: forc-plugins/forc-tx/src/lib.rs (hunk 4 above)
Comment (xgreenx) [reply forc-plugins/forc-tx/src/lib.rs:70]: We don't need to support `Mint` transactions on the `forc` level. Because it is not allowed to send transactions like this. It is only produced by block producer

File: forc-plugins/forc-client/src/util/gas.rs (hunk 5 above)
Comment (xgreenx) [reply forc-plugins/forc-client/src/util/gas.rs:20]: Why can't we reuse the logic from Rust SDK to estimate/send transactions?

File: forc-plugins/forc-tx/src/lib.rs (hunk 4 above)
Comment (hal3e) [reply forc-plugins/forc-tx/src/lib.rs:70]: I removed the `Mint` variant

File: forc-plugins/forc-client/src/util/gas.rs (hunk 5 above)
Comment (hal3e) [reply forc-plugins/forc-client/src/util/gas.rs]: In this case we can as we have a provider. I have updated the code.

File: Cargo.toml [hunk 8]
```diff
@@ -33,21 +33,22 @@ exclude = [
...
 
 # Dependencies from the `fuels-rs` repository:
-fuels-core = "0.45"
-fuels-accounts = "0.45"
+fuels-core = "0.53.0"
+fuels-accounts = "0.53.0"
 
 # Dependencies from the `forc-wallet` repository:
-forc-wallet = "0.2.5"
+forc-wallet = "0.3.0"
```
Comment (kayagokalp) [inline Cargo.toml:51]:
```suggestion
forc-wallet = "0.4.0"
```

File: Cargo.toml [hunk 9]
```diff
@@ -56,3 +57,6 @@ homepage = "https://fuel.network/"
 license = "Apache-2.0"
 repository = "https://github.com/FuelLabs/sway"
 
+[patch.crates-io]
+# Keep it until release of the `forc-wallet`.
+forc-wallet = { git = "https://github.com/FuelLabs/forc-wallet", branch = "chore/bump-fuels-rs-fuel-core" }
```
Comment (kayagokalp) [inline Cargo.toml:62]:
```suggestion
```

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -255,7 +255,7 @@ jobs:
     runs-on: ubuntu-latest
     services:
       fuel-core:
-        image: ghcr.io/fuellabs/fuel-core:v0.20.3
+        image: ghcr.io/fuellabs/fuel-core:latest
```
Comment (JoshuaBatty) [inline .github/workflows/ci.yml:258]:
```suggestion
        image: ghcr.io/fuellabs/fuel-core:v0.21.0
```

File: forc-plugins/forc-tx/src/lib.rs [hunk 11]
```diff
@@ -668,15 +659,19 @@ impl TryFrom<Script> for fuel_tx::Script {
...
-            script
-                .gas
-                .limit
-                .unwrap_or(fuel_tx::ConsensusParameters::DEFAULT.max_gas_per_tx),
-            script.maturity.maturity.into(),
+            script.gas.script_gas_limit.unwrap_or(
+                fuel_tx::ConsensusParameters::default()
+                    .tx_params()
+                    .max_gas_per_tx
+                    / 4,
```
Comment (JoshuaBatty) [inline forc-plugins/forc-tx/src/lib.rs:670]: Is this correct to divide by 4? If so, should there be another field we can use? Or at the very least a comment explaining what the logic here is? It's not clear to me from reading the code why this value is used. cc @xgreenx

File: forc-test/src/lib.rs [hunk 12]
```diff
@@ -793,7 +796,12 @@ fn exec_test(
             tx_pointer,
             0u32.into(),
         )
-        .gas_limit(tx::ConsensusParameters::DEFAULT.max_gas_per_tx)
+        .script_gas_limit(
+            tx::ConsensusParameters::default()
+                .tx_params()
+                .max_gas_per_tx
+                / 4,
```
Comment (JoshuaBatty) [inline forc-test/src/lib.rs:803]: same here

File: sway-lsp/tests/fixtures/benchmark/Forc.lock
Comment (JoshuaBatty) [inline sway-lsp/tests/fixtures/benchmark/Forc.lock:1]: We don't need this file checked in, can we delete this file?

File: test/src/e2e_vm_tests/harness.rs [hunk 13]
```diff
@@ -149,22 +149,25 @@ pub(crate) fn runs_in_vm(
...
-                    rng.gen(),
+                    SecretKey::random(rng),
                     rng.gen(),
                     1,
                     Default::default(),
                     rng.gen(),
                     0u32.into(),
                 )
-                .gas_limit(fuel_tx::ConsensusParameters::DEFAULT.max_gas_per_tx)
+                .script_gas_limit(ConsensusParameters::default().tx_params().max_gas_per_tx / 4)
```
Comment (JoshuaBatty) [inline test/src/e2e_vm_tests/harness.rs:170]: Here is this magic number 4 again....

**Review by JoshuaBatty — CHANGES_REQUESTED** [review]

File: sway-lib-std/src/inputs.sw [hunk 14]
```diff
@@ -19,44 +19,44 @@ use ::tx::{
...
+// pub const GTF_INPUT_CONTRACT_TX_POINTER = 0x224;
+// pub const GTF_INPUT_CONTRACT_CONTRACT_ID = 0x225;
+pub const GTF_INPUT_MESSAGE_SENDER = 0x240;
+pub const GTF_INPUT_MESSAGE_RECIPIENT = 0x241;
+pub const GTF_INPUT_MESSAGE_AMOUNT = 0x242;
+pub const GTF_INPUT_MESSAGE_NONCE = 0x243;
 // These are based on the old spec (before
 // https://github.com/FuelLabs/fuel-specs/pull/400) because that's what's
 // currently implemented in `fuel-core`, `fuel-asm`, and `fuel-tx. They should
 // eventually be updated.
```
Comment (IGI-111) [inline sway-lib-std/src/inputs.sw:52]: Is the comment still relevant since we are updating them?

File: sway-lib-std/src/tx.sw [hunk 15]
```diff
@@ -70,9 +72,6 @@ pub enum Transaction {
 ///         Transaction::Script => {
 ///             log("Regular script transaction");
 ///         },
-///         Transaction::Create => {
-///             log("Contract deployment transaction");
-///         },
```
Comment (IGI-111) [inline sway-lib-std/src/tx.sw:75]: `Create` still exists as a variant, so why is this removed?

Comment (Voxelot) [reply sway-lib-std/src/tx.sw:75]: I think it's because the only time a creation tx would be running any sway code would be inside of predicates, and predicates can't log.

File: sway-lib-std/src/inputs.sw (hunk 14 above)
Comment (hal3e) [reply sway-lib-std/src/inputs.sw:52]: Not anymore. Deleted the comments.

**Review by JoshuaBatty — APPROVED** [review]

**Review by IGI-111 — APPROVED** [review]
